PR: https://github.com/odoo/odoo/pull/235832

From: 3078021ca336f1e090891285f49e6ecf4d38a31a
From: Krzysztof Magusiak (krma)
Date: 2026-01-22 23:51:10

Breaking data model changes scores: del:14 + add:0, change matches:
-    needed_terms = fields.Binary(compute='_compute_needed_terms', exportable=False)
-    invoice_outstanding_credits_debits_widget = fields.Binary(
-    invoice_payments_widget = fields.Binary(
-    tax_totals = fields.Binary(
-    payment_term_details = fields.Binary(compute="_compute_payment_term_details", exportable=False)
-    term_key = fields.Binary(compute='_compute_term_key', exportable=False)
-    epd_key = fields.Binary(compute='_compute_epd_key', exportable=False)
-    epd_needed = fields.Binary(compute='_compute_epd_needed', exportable=False)
-    discount_allocation_key = fields.Binary(compute='_compute_discount_allocation_key', exportable=False)
-    discount_allocation_needed = fields.Binary(compute='_compute_discount_allocation_needed', exportable=False)
-    tag_ids_domain = fields.Binary(string="tag domain", help="Dynamic domain used for the tag that can be set on tax", compute="_compute_tag_ids_domain")
-    account_templates = fields.Binary(compute='_compute_account_templates', exportable=False)
-    account_map = fields.Binary(compute='_compute_account_map')
-    tax_map = fields.Binary(compute='_compute_tax_map')

Total Changes: 100

[IMP] orm, *: fields.Binary value is bytes

Stop accepting any kind of values in binary fields. This was used as a
hack to store mutable data inside the cache of the field.

Some models use this hack to store arbitrary data. They should use the
correct data type (mostly fields.Json) or eventually the transaction
cache.

task-4251301

closes odoo/odoo#235832

Related: odoo/enterprise#99567
Related: odoo/upgrade#9314
Signed-off-by: Raphael Collet <rco@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -406,7 +406,7 @@ class AccountMove(models.Model):
         inverse='_inverse_invoice_payment_term_id',
         check_company=True,
     )
-    needed_terms = fields.Binary(compute='_compute_needed_terms', exportable=False)
+    needed_terms = fields.Json(compute='_compute_needed_terms', exportable=False)
     needed_terms_dirty = fields.Boolean(compute='_compute_needed_terms')
     tax_calculation_rounding_method = fields.Selection(
         related='company_id.tax_calculation_rounding_method',
@@ -486,7 +486,7 @@ class AccountMove(models.Model):
     )
 
     # === Payment widget fields === #
-    invoice_outstanding_credits_debits_widget = fields.Binary(
+    invoice_outstanding_credits_debits_widget = fields.Json(
         groups="account.group_account_invoice,account.group_account_readonly",
         compute='_compute_payments_widget_to_reconcile_info',
         exportable=False,
@@ -495,7 +495,7 @@ class AccountMove(models.Model):
         groups="account.group_account_invoice,account.group_account_readonly",
         compute='_compute_invoice_has_outstanding',
     )
-    invoice_payments_widget = fields.Binary(
+    invoice_payments_widget = fields.Json(
         groups="account.group_account_invoice,account.group_account_readonly",
         compute='_compute_payments_widget_reconciled_info',
         exportable=False,
@@ -587,7 +587,7 @@ class AccountMove(models.Model):
         compute='_compute_amount', store=True,
         currency_field='company_currency_id',
     )
-    tax_totals = fields.Binary(
+    tax_totals = fields.Json(
         string="Invoice Totals",
         compute='_compute_tax_totals',
         inverse='_inverse_tax_totals',
@@ -757,7 +757,7 @@ class AccountMove(models.Model):
     show_update_fpos = fields.Boolean(string="Has Fiscal Position Changed", store=False)  # True if the fiscal position was changed
 
     # used to display the various dates and amount dues on the invoice's PDF
-    payment_term_details = fields.Binary(compute="_compute_payment_term_details", exportable=False)
+    payment_term_details = fields.Json(compute="_compute_payment_term_details", exportable=False)
     show_payment_term_details = fields.Boolean(compute="_compute_show_payment_term_details")
     show_discount_details = fields.Boolean(compute="_compute_show_payment_term_details")
 
@@ -1072,10 +1072,9 @@ class AccountMove(models.Model):
     def _compute_invoice_date_due(self):
         today = fields.Date.context_today(self)
         for move in self:
-            move.invoice_date_due = move.needed_terms and max(
-                (k['date_maturity'] for k in move.needed_terms.keys() if k),
-                default=False,
-            ) or move.invoice_date_due or today
+            needed_terms = move.needed_terms or ()
+            due_date = max((k['date_maturity'] for k, _ in needed_terms if k), default=False)
+            move.invoice_date_due = due_date or move.invoice_date_due or today
 
     def _compute_delivery_date(self):
         pass
@@ -1359,7 +1358,7 @@ class AccountMove(models.Model):
         AccountTax = self.env['account.tax']
         for invoice in self.with_context(bin_size=False):
             is_draft = invoice.id != invoice._origin.id
-            invoice.needed_terms = {}
+            needed_terms = {}
             invoice.needed_terms_dirty = True
             sign = 1 if invoice.is_inbound(include_receipts=True) else -1
             if invoice.is_invoice(True) and invoice.invoice_line_ids:
@@ -1408,13 +1407,13 @@ class AccountMove(models.Model):
                             'discount_balance': invoice_payment_terms.get('discount_balance') or 0.0,
                             'discount_amount_currency': invoice_payment_terms.get('discount_amount_currency') or 0.0,
                         }
-                        if key not in invoice.needed_terms:
-                            invoice.needed_terms[key] = values
+                        if key not in needed_terms:
+                            needed_terms[key] = values
                         else:
-                            invoice.needed_terms[key]['balance'] += values['balance']
-                            invoice.needed_terms[key]['amount_currency'] += values['amount_currency']
+                            needed_terms[key]['balance'] += values['balance']
+                            needed_terms[key]['amount_currency'] += values['amount_currency']
                 else:
-                    invoice.needed_terms[frozendict({
+                    needed_terms[frozendict({
                         'move_id': invoice.id,
                         'date_maturity': fields.Date.to_date(invoice.invoice_date_due),
                         'discount_date': False,
@@ -1424,6 +1423,7 @@ class AccountMove(models.Model):
                         'balance': invoice.amount_total_signed,
                         'amount_currency': invoice.amount_total_in_currency_signed,
                     }
+            invoice.needed_terms = list(needed_terms.items())
 
     @api.depends('suitable_journal_ids')
     def _compute_show_journal(self):
@@ -1798,18 +1798,19 @@ class AccountMove(models.Model):
         for move in self:
             if move.is_invoice(include_receipts=True):
                 base_lines, _tax_lines = move._get_rounded_base_and_tax_lines()
-                move.tax_totals = self.env['account.tax']._get_tax_totals_summary(
+                tax_totals = self.env['account.tax']._get_tax_totals_summary(
                     base_lines=base_lines,
                     currency=move.currency_id,
                     company=move.company_id,
                     cash_rounding=move.invoice_cash_rounding_id,
                 )
-                move.tax_totals['display_in_company_currency'] = (
+                tax_totals['display_in_company_currency'] = (
                     move.company_id.display_invoice_tax_company_currency
                     and move.company_currency_id != move.currency_id
-                    and move.tax_totals['has_tax_groups']
+                    and tax_totals['has_tax_groups']
                     and move.is_sale_document(include_receipts=True)
                 )
+                move.tax_totals = tax_totals
             else:
                 # Non-invoice moves don't support that field (because of multicurrency: all lines of the invoice share the same currency)
                 move.tax_totals = None
@@ -3197,9 +3198,10 @@ class AccountMove(models.Model):
     def _sync_dynamic_line_needed_values(self, values_list):
         res = {}
         for computed_needed in values_list:
-            if computed_needed is False:
+            if not computed_needed:
                 continue  # there was an invalidation, let's hope nothing needed to be changed...
-            for key, values in computed_needed.items():
+            for key, values in computed_needed:
+                key = frozendict(key)
                 if key not in res:
                     res[key] = dict(values)
                 else:
@@ -3528,9 +3530,9 @@ class AccountMove(models.Model):
     def _sync_dynamic_line(self, existing_key_fname, needed_vals_fname, needed_dirty_fname, line_type, container):
         def existing():
             return {
-                line: line[existing_key_fname]
+                line: frozendict(line_value)
                 for line in container['records'].line_ids
-                if line[existing_key_fname]
+                if (line_value := line[existing_key_fname])
             }
 
         def needed():

--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -421,12 +421,12 @@ class AccountMoveLine(models.Model):
     deductible_amount = fields.Float("Deductibility", default=100)
 
     # === Invoice sync fields === #
-    term_key = fields.Binary(compute='_compute_term_key', exportable=False)
-    epd_key = fields.Binary(compute='_compute_epd_key', exportable=False)
-    epd_needed = fields.Binary(compute='_compute_epd_needed', exportable=False)
+    term_key = fields.Json(compute='_compute_term_key', exportable=False)
+    epd_key = fields.Json(compute='_compute_epd_key', exportable=False)
+    epd_needed = fields.Json(compute='_compute_epd_needed', exportable=False)
     epd_dirty = fields.Boolean(compute='_compute_epd_needed')
-    discount_allocation_key = fields.Binary(compute='_compute_discount_allocation_key', exportable=False)
-    discount_allocation_needed = fields.Binary(compute='_compute_discount_allocation_needed', exportable=False)
+    discount_allocation_key = fields.Json(compute='_compute_discount_allocation_key', exportable=False)
+    discount_allocation_needed = fields.Json(compute='_compute_discount_allocation_needed', exportable=False)
     discount_allocation_dirty = fields.Boolean(compute='_compute_discount_allocation_needed')
 
     # === Analytic fields === #
@@ -1094,7 +1094,12 @@ class AccountMoveLine(models.Model):
                 })
                 dist = distribution_totals[key]
                 total = sum(dist.values()) or 1  # avoid division by zero
-                discount_allocation_needed[key] = frozendict({
+                key_needed = frozendict({
+                    'move_id': line.move_id._origin.id,
+                    'account_id': account._origin.id,
+                    'currency_rate': line.currency_rate,
+                }) if not line.move_id.id else key
+                discount_allocation_needed[key_needed] = frozendict({
                     'display_type': 'discount',
                     'name': _("Discount"),
                     'amount_currency': amount,
@@ -1103,7 +1108,7 @@ class AccountMoveLine(models.Model):
                         for account_id, value in dist.items()
                     }
                 })
-            line.discount_allocation_needed = discount_allocation_needed
+            line.discount_allocation_needed = list(discount_allocation_needed.items())
 
     @api.depends('tax_ids', 'account_id', 'company_id')
     def _compute_epd_key(self):
@@ -1141,8 +1146,8 @@ class AccountMoveLine(models.Model):
             taxes = line.tax_ids.filtered(lambda t: t.amount_type != 'fixed')
             epd_needed_vals = epd_needed.setdefault(
                 frozendict({
-                    'move_id': line.move_id.id,
-                    'account_id': line.account_id.id,
+                    'move_id': line.move_id._origin.id,
+                    'account_id': line.account_id._origin.id,
                     'analytic_distribution': line.analytic_distribution,
                     'tax_ids': [Command.set(taxes.ids)],
                     'display_type': 'epd',
@@ -1161,8 +1166,8 @@ class AccountMoveLine(models.Model):
             epd_needed_vals['balance'] -= balance
             epd_needed_vals = epd_needed.setdefault(
                 frozendict({
-                    'move_id': line.move_id.id,
-                    'account_id': line.account_id.id,
+                    'move_id': line.move_id._origin.id,
+                    'account_id': line.account_id._origin.id,
                     'display_type': 'epd',
                 }),
                 {
@@ -1174,7 +1179,7 @@ class AccountMoveLine(models.Model):
             )
             epd_needed_vals['amount_currency'] += amount_currency
             epd_needed_vals['balance'] += balance
-            line.epd_needed = {k: frozendict(v) for k, v in epd_needed.items()}
+            line.epd_needed = list(epd_needed.items())
 
     @api.depends('move_id.move_type', 'balance', 'tax_repartition_line_id', 'tax_ids')
     def _compute_is_refund(self):
@@ -1205,7 +1210,7 @@ class AccountMoveLine(models.Model):
         for line in self:
             if line.display_type == 'payment_term':
                 line.term_key = frozendict({
-                    'move_id': line.move_id.id,
+                    'move_id': line.move_id._origin.id,
                     'date_maturity': fields.Date.to_date(line.date_maturity),
                     'discount_date': line.discount_date,
                 })

--- a/addons/account/models/account_tax.py
+++ b/addons/account/models/account_tax.py
@@ -4952,7 +4952,7 @@ class AccountTaxRepartitionLine(models.Model):
         compute='_compute_use_in_tax_closing', store=True, readonly=False, precompute=True,
     )
 
-    tag_ids_domain = fields.Binary(string="tag domain", help="Dynamic domain used for the tag that can be set on tax", compute="_compute_tag_ids_domain")
+    tag_ids_domain = fields.Json(string="tag domain", help="Dynamic domain used for the tag that can be set on tax", compute="_compute_tag_ids_domain")
 
     @api.depends('company_id.multi_vat_foreign_country_ids', 'company_id.account_fiscal_country_id')
     def _compute_tag_ids_domain(self):

--- a/addons/account/models/chart_template.py
+++ b/addons/account/models/chart_template.py
@@ -121,6 +121,7 @@ class AccountChartTemplate(models.AbstractModel):
         return {
             name: template
             for mapping in modules.mapped('account_templates')
+            if mapping
             for name, template in mapping.items()
             if get_all or template['visible']
         }

--- a/addons/account/models/ir_module.py
+++ b/addons/account/models/ir_module.py
@@ -24,7 +24,7 @@ TEMPLATE_REGISTER = {}
 class IrModuleModule(models.Model):
     _inherit = "ir.module.module"
 
-    account_templates = fields.Binary(compute='_compute_account_templates', exportable=False)
+    account_templates = fields.Json(compute='_compute_account_templates', exportable=False)
 
     @api.depends('state')
     def _compute_account_templates(self):
@@ -119,7 +119,12 @@ class IrModuleModule(models.Model):
             del self.env.registry._auto_install_template
 
     def module_uninstall(self):
-        unlinked_templates = [code for template in self.mapped('account_templates') for code in template]
+        unlinked_templates = [
+            code
+            for template in self.mapped('account_templates')
+            if template
+            for code in template
+        ]
         if unlinked_templates:
             companies = self.env['res.company'].search([
                 ('chart_template', 'in', unlinked_templates),

--- a/addons/account/models/partner.py
+++ b/addons/account/models/partner.py
@@ -53,7 +53,7 @@ class AccountFiscalPosition(models.Model):
         string='Company', required=True, readonly=True, index=True,
         default=lambda self: self.env.company)
     account_ids = fields.One2many('account.fiscal.position.account', 'position_id', string='Account Mapping', copy=True)
-    account_map = fields.Binary(compute='_compute_account_map')
+    account_map = fields.Json(compute='_compute_account_map')
     tax_ids = fields.Many2many(
         comodel_name='account.tax',
         relation='account_fiscal_position_account_tax_rel',
@@ -61,7 +61,7 @@ class AccountFiscalPosition(models.Model):
         column2='account_tax_id',
         string='Taxes',
     )
-    tax_map = fields.Binary(compute='_compute_tax_map')
+    tax_map = fields.Json(compute='_compute_tax_map')
     note = fields.Html('Notes', translate=True, help="Legal mentions that have to be printed on the invoices.")
     auto_apply = fields.Boolean(string='Detect Automatically', help="Apply tax & account mappings on invoices automatically if the matching criterias (VAT/Country) are met.")
     vat_required = fields.Boolean(string='VAT required', help="Apply only if partner has a VAT number.")
@@ -171,11 +171,12 @@ class AccountFiscalPosition(models.Model):
         self.ensure_one()
         if not self.tax_ids:  # empty fiscal positions (like those created by tax units) remove all taxes
             return self.env['account.tax']
+        tax_map = self.tax_map or {}
         return self.env['account.tax'].browse(unique(
             tax_id
             for tax in taxes
-            for tax_id in self.tax_map.get(
-                tax.id,
+            for tax_id in tax_map.get(
+                str(tax.id),
                 # If not in tax_map, a tax is mapped to itself if it has 'self' as fiscal position
                 # or it has no fiscal position. Else it's removed.
                 [tax.id] if self in tax.fiscal_position_ids or not tax.fiscal_position_ids else [],
@@ -183,7 +184,7 @@ class AccountFiscalPosition(models.Model):
         ))
 
     def map_account(self, account):
-        return self.env['account.account'].browse((self.account_map or {}).get(account.id, account.id))
+        return self.env['account.account'].browse((self.account_map or {}).get(str(account.id), account.id))
 
     @api.onchange('country_id')
     def _onchange_country_id(self):
